pkg/util/strategicpatch/patch.go

```diff
@@ -0,0 +1,454 @@
+//
+// For more information, see the PATCH section of docs/api-conventions.md.
+func StrategicMergePatchData(original, patch []byte, dataStruct interface{}) ([]byte, error) {
+	var o map[string]interface{}
+	err := json.Unmarshal(original, &o)
+	if err != nil {
+		return nil, err
+	}
+
+	var p map[string]interface{}
+	err = json.Unmarshal(patch, &p)
+	if err != nil {
+		return nil, err
+	}
+
+	t := reflect.TypeOf(dataStruct)
+	if t.Kind() == reflect.Ptr {
+		t = t.Elem()
+	}
+	if t.Kind() != reflect.Struct {
+		return nil, fmt.Errorf("strategic merge patch needs a struct, %s received instead", t.Kind().String())
+	}
+
+	result, err := mergeMap(o, p, t)
+	if err != nil {
+		return nil, err
+	}
+
+	return json.Marshal(result)
+}
+
+const specialKey = "__patch"
+
+// Merge fields from a patch map into the original map.
+func mergeMap(original, patch map[string]interface{}, t reflect.Type) (map[string]interface{}, error) {
+	// If the map contains "__patch: strict", don't merge it, just use the
+	// patch map directly. Later on, can add a non-recursive strict that only
+	// affects the map that the __patch is in.
+	if v, ok := patch[specialKey]; ok {
+		if v == "strict" {
+			delete(patch, specialKey)
+			return patch, nil
+		}
+	}
+
+	// Start merging the patch into the original.
+	for k, patchV := range patch {
+		// If the value of this key is null, delete the key if it exists in the
+		// original. Otherwise, skip it.
+		if patchV == nil {
+			if _, ok := original[k]; ok {
+				delete(original, k)
+			}
+			continue
+		}
+
+		_, ok := original[k]
+		if !ok {
+			// If it's not in the original document, just take the patch value.
+			if original == nil {
+				original = map[string]interface{}{}
+			}
+			original[k] = patchV
+		} else {
+			// If they're both maps or lists, recurse into the value.
+			// First find the fieldPatchStrategy and fieldPatchMergeKey.
+			fieldType, fieldPatchStrategy, fieldPatchMergeKey, err := lookupPatchMetadata(t, k)
+			if err != nil {
+				return nil, err
+			}
+
+			originalType := reflect.TypeOf(original[k])
+			patchType := reflect.TypeOf(patchV)
+			if originalType == patchType {
+				if originalType.Kind() == reflect.Map && fieldPatchStrategy != "replace" {
+					typedOriginal := original[k].(map[string]interface{})
+					typedPatch := patchV.(map[string]interface{})
+					var err error
+					original[k], err = mergeMap(typedOriginal, typedPatch, fieldType)
+					if err != nil {
+						return nil, err
+					}
+					continue
+				} else if originalType.Kind() == reflect.Slice && fieldPatchStrategy == "merge" {
+					elemType := fieldType.Elem()
+					typedOriginal := original[k].([]interface{})
+					typedPatch := patchV.([]interface{})
+					var err error
+					original[k], err = mergeSlice(typedOriginal, typedPatch, elemType, fieldPatchMergeKey)
+					if err != nil {
+						return nil, err
+					}
+					continue
+				}
+			}
+
+			original[k] = patchV
+		}
+	}
+
+	return original, nil
```

Comment:
We need to mutate because deep-copying a slice or map in golang to then work on the copy is very difficult. (I added a comment to the function to clarify this.)

I tried just now to refactor to only mutate and not return, but sometimes we just return the patch value directly (e.g. in the "strict" case) or generate a new value, and trying to mutate the original in that case would add a good amount of complexity. Hence, both mutating and returning. Given these are private functions with very specific use, I think it's not great but okay.